File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/actions/Action.java
package info.nightscout.androidaps.plugins.general.automation.actions;

import android.widget.LinearLayout;

import com.google.common.base.Optional;

import org.json.JSONException;
import org.json.JSONObject;

import javax.annotation.Nullable;

import info.nightscout.androidaps.plugins.general.automation.triggers.Trigger;
import info.nightscout.androidaps.queue.Callback;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/*
    Action ideas:

    * cancel temp target
    * change preference setting
    * enable/disable plugin
    * create notification
    * create ugly alarm
    * create profile switch
    * set/cancel tbr
    * set/cancel extended bolus
    * run bolus wizard

    Trigger ideas:

    * location (close to)
    * connected to specific wifi
    * internet available/not available
    * nsclient connected/disconnected
    * iob
    * cob
    * autosens value
    * delta, short delta, long delta
    * last bolus ago
    * is tbr running
    * bolus wizard result
    * loop is enabled, disabled, suspended, running

*/


public abstract class Action {
    private static final Logger log = LoggerFactory.getLogger(Action.class);

    public Trigger precondition = null;

    public abstract int friendlyName();

    public abstract String shortDescription();

    public abstract void doAction(Callback callback);

    public void generateDialog(LinearLayout root) {
    }

    public boolean hasDialog() {
        return false;
    }

    public String toJSON() {
        JSONObject o = new JSONObject();
        try {
            o.put("type", this.getClass().getName());
        } catch (JSONException e) {
            log.error("Unhandled exception", e);
        }
        return o.toString();
    }

    public abstract Optional<Integer> icon();

    public Action fromJSON(String data) {
        return this;
    }

    @Nullable
    public static Action instantiate(JSONObject object) {
        try {
            String type = object.getString("type");
            JSONObject data = object.optJSONObject("data");
            Class clazz = Class.forName(type);
            return ((Action) clazz.newInstance()).fromJSON(data != null ? data.toString() : "");
        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | JSONException e) {
            log.error("Unhandled exception", e);
        }
        return null;
    }

    public void apply(Action a) {
        try {
            JSONObject object = new JSONObject(a.toJSON());
            String type = object.getString("type");
            JSONObject data = object.getJSONObject("data");
            if (type.equals(getClass().getName())) {
                fromJSON(data.toString());
            }
        } catch (JSONException e) {
            log.error("Unhandled exception", e);
        }
    }
}

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/actions/Action.kt
package info.nightscout.androidaps.plugins.general.automation.actions

import android.widget.LinearLayout
import androidx.annotation.DrawableRes
import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.logging.AAPSLogger
import info.nightscout.androidaps.plugins.general.automation.triggers.Trigger
import info.nightscout.androidaps.queue.Callback
import org.json.JSONException
import org.json.JSONObject
import javax.inject.Inject
import kotlin.reflect.full.primaryConstructor

abstract class Action(val injector: HasAndroidInjector) {

    @Inject lateinit var aapsLogger: AAPSLogger

    var precondition: Trigger? = null

    abstract fun friendlyName(): Int
    abstract fun shortDescription(): String
    abstract fun doAction(callback: Callback)
    abstract fun isValid(): Boolean
    @DrawableRes abstract fun icon(): Int

    init {
        injector.androidInjector().inject(this)
    }

    open fun generateDialog(root: LinearLayout) {}

    open fun hasDialog(): Boolean = false

    open fun toJSON(): String =
        JSONObject().put("type", this.javaClass.name).toString()

    open fun fromJSON(data: String): Action = this

    fun apply(a: Action) {
        val obj = JSONObject(a.toJSON())
        val type = obj.getString("type")
        val data = obj.getJSONObject("data")
        if (type == javaClass.name) fromJSON(data.toString())
    }

    fun instantiate(obj: JSONObject): Action? {
        try {
            val type = obj.getString("type")
            val data = obj.optJSONObject("data")
            val clazz = Class.forName(type).kotlin
            return (clazz.primaryConstructor?.call(injector) as Action).fromJSON(data?.toString()
                ?: "")
            //return (clazz.newInstance() as Action).fromJSON(data?.toString() ?: "")
        } catch (e: ClassNotFoundException) {
            aapsLogger.error("Unhandled exception", e)
        } catch (e: InstantiationException) {
            aapsLogger.error("Unhandled exception", e)
        } catch (e: IllegalAccessException) {
            aapsLogger.error("Unhandled exception", e)
        } catch (e: JSONException) {
            aapsLogger.error("Unhandled exception", e)
        }
        return null
    }
}